src/importer.rs

```diff
@@ -0,0 +1,284 @@
+        let checksum = calculate_checksum(&file_contents);
+        let is_identical_obj_loaded =
+            self.checksum_paths.contains_key(&checksum) && !self.loaded_models.contains_key(path);
+
+        // Identical models imported from different path are duplicated.
+        let models = if is_identical_obj_loaded {
+            self.duplicate_models(checksum, path);
+
+            self.loaded_models
+                .get(path)
+                .expect("Model should be present in cache")
+                .clone()
+        } else {
+            match self.path_checksums.entry(path.to_string()) {
+                // This file was already imported and we need to check its
+                // checksum and either just return parsed models or reimport
+                // and recache it.
+                Entry::Occupied(mut path_checksum) => {
+                    if checksum == *path_checksum.get() {
+                        self.loaded_models
+                            .get(path)
+                            .expect("Model should be present in cache")
+                            .clone()
+                    } else {
+                        match obj_buf_into_tobj(&file_contents) {
+                            Ok((tobj_models, _)) => {
+                                let models = tobj_to_internal(tobj_models);
+                                let old_checksum = *path_checksum.get();
+
+                                self.loaded_models.insert(path.to_string(), models.clone());
+                                path_checksum.insert(checksum);
+                                self.remove_checksum_path(old_checksum, &path);
+                                self.add_checksum_path(checksum, &path);
+
+                                models
+                            }
+                            Err(_) => return Err(ImporterError::InvalidStructure),
+                        }
+                    }
+                }
+                // This is a brand new file to parse.
+                Entry::Vacant(path_checksum) => match obj_buf_into_tobj(&file_contents) {
+                    Ok((tobj_models, _)) => {
+                        let models = tobj_to_internal(tobj_models);
+
+                        self.loaded_models.insert(path.to_string(), models.clone());
+                        path_checksum.insert(checksum);
+                        self.checksum_paths.insert(checksum, vec![path.to_string()]);
+
+                        models
+                    }
+                    Err(_) => return Err(ImporterError::InvalidStructure),
+                },
+            }
+        };
+
+        Ok(models)
+    }
+
+    /// Loads contents of file to string. In case of any error, it is converted
+    /// into `ImporterError`.
+    fn load_file_to_string(&self, path: &str) -> Result<String, ImporterError> {
+        match fs::read_to_string(path) {
+            Ok(contents) => Ok(contents),
+            Err(err) => match err.kind() {
```

Comment:
This can be a `impl From<io::Error> for ImporterError` and this whole function just disappears. `?` operator will do the conversion for you once you have `From`.

Reply:
Done.